lib/uxtx.js

```diff
@@ -0,0 +1,307 @@
+const moment = require('moment');
+const assert = require('bsert');
+
+const { Amount, TX } = require('bcoin');
+
+// TODO: rename this.getJSON to
+// this.getTXJSON
+// TODO: value to BTC, prepend + or -
+
+
+// bcoin api
+// deposits:
+// null path in the inputs
+// output with path defined and change false
+//
+// withdrawals:
+// input with defined path and change false
+// output with path defined and change true
+
+class TXUX extends TX {
+  constructor(options) {
+    super();
+
+    this.TYPES = {
+      coinbase: 'coinbase',
+      withdraw: 'withdraw',
+      deposit: 'deposit',
+      unknown: 'unknown'
+    }
+
+    this.dateFormat = null;
+    this._UXType = null;
+    this._copy = null;
+    this._json = null;
+    this._wallet = null;
+    this._account = null;
+    this._counterparty = null;
+    this._recipients = null;
+
+    if (options)
+      this.fromOptions(options);
+  }
+
+  fromOptions(options) {
+    if (options.dateFormat) {
+      assert(typeof options.dateFormat === 'string');
+      // TODO: figure out correct date format string
+      //assert(moment(options.dateFormat).isValid());
+      this.dateFormat = options.dateFormat;
+    }
+    if (options.copy) {
+      assert(typeof options.copy === 'object');
+      for (let [key, val] of Object.entries(options.copy))
+        assert(typeof val === 'string');
+      this._copy = options.copy;
+    }
+    if (options.json) {
+      assert(typeof options.json === 'object');
+      this._json = options.json;
+    }
+
+    return this;
+  }
+
+  /*
+   * withdrawal defined by:
+   * more than 1 known input
+   * at least 1 change output
+   * TODO: not all withdrawals have change
+   */
+  isWithdraw() {
+    const json = this.getJSON();
+
+    const knownOutputs = this.getKnownCoins('outputs');
+    const knownInputs = this.getKnownCoins('inputs');
+    const changeOutputs = this.getChangeCoins(knownOutputs);
+
+    if (knownInputs.length > 0 && changeOutputs.length > 0)
+      return true;
+      //sendOrReceive = constants.SEND;
+
+    return false;
+  }
+
+
+  /*
+   * deposit defined by:
+   * no known inputs
+   * no change outputs
+   */
+  isDeposit() {
+    const json = this.getJSON();
+    const knownInputs = this.getKnownCoins('inputs');
+    const changeOutputs = this.getChangeCoins(json.outputs);
+
+    if (knownInputs.length === 0 && changeOutputs.length === 0)
+      return true;
+      //sendOrReceive = constants.RECEIVE;
+
+    return false;
+  }
+
+  // type can be input, outputs
+  // undefined means both
+  // TODO: cache and partially cache
+  getKnownCoins(type) {
+    const json = this.getJSON();
+
+    if (type === 'inputs')
+      return json.inputs.filter(i => i.path);
+    if (type === 'outputs')
+      return json.outputs.filter(o => o.path);
+
+    return [
```

Comment:
Would this be more useful as an object? That way you could destructure out what you need and don't have to know the order in order to extract.

Reply:
That would make the return value polymorphic, sometimes we want all of the coins

Reply:
Ah ok, I misread that as a tuple. Sorry! Makes sense 👍

Reply:
I can add a comment around this to make it more clear, misreading I think means its not clear enough 👍